### PR TITLE
_1password: 1.12.2 -> 2.0.0

### PR DESCRIPTION
**Pull request by malob**
###### Description of changes

Update 1Password CLI to version 2.0.0

I also added a second commit (b60edec46269914df515aaa35b0632280cefdec1) that refactors the code to reduce some duplication. Happy to remove it if it's overly fancy etc., but I like it :)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by malob** [comment]
Result of `nixpkgs-review pr 165321` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_1password</li>
  </ul>
</details>

**Comment by malob** [comment]
Result of `nixpkgs-review pr 165321` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_1password</li>
  </ul>
</details>

**Comment by malob** [comment]
Result of `nixpkgs-review pr 165321` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_1password</li>
  </ul>
</details>

File: pkgs/applications/misc/1password/default.nix [hunk 1]
```diff
@@ -1,28 +1,27 @@
...
-        sha256 = "xG/6YZdkJxr5Py90rkIyG4mK40yFTmNSfih9jO2uF+4=";
-      };
+  inherit version;
+  src = rec {
+    aarch64-linux = fetch "linux_arm64" "sha256-NhCs68on8LzoeOmM5eP8LwmFaVWz6aghqtHzfUlACiA=" "zip";
+    i686-linux = fetch "linux_386" "sha256-vCxgEBq4YVfljq2zUpvBdZUbIiam4z64P1m9OMWq1f4=" "zip";
+    x86_64-linux = fetch "linux_amd64" "sha256-CDwrJ5ksXf9kwHobw4jvRUi1hLQzq4/yRlk+kHPN7UE=" "zip";
+    aarch64-darwin = fetch "apple_universal" "sha256-DC9hdzRjQ9iNjbe6PfRpMXzDeInq4rYSAa2nDHQMTRo=" "pkg";
+    x86_64-darwin = aarch64-darwin;
+  }.${stdenv.hostPlatform.system};
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/1password/default.nix:24]: Missing or throw unsupported platform.

Comment (malob) [reply pkgs/applications/misc/1password/default.nix:24]: @SuperSandro2000, is there an issue here? There's a source for each of the listed supported platforms:
```
platforms = [ "i686-linux" "x86_64-linux" "x86_64-darwin" "aarch64-linux" "aarch64-darwin" ];
```

Should be fine as long as that's true right? Maybe I'm missing something?

**Comment by maxeaubrey** [comment]
I can't comment on the other comment, but could you add installing shell completions please? https://developer.1password.com/docs/cli/reference/commands/completion/

File: pkgs/applications/misc/1password/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/applications/misc/1password/default.nix:24]: Yes, you would get a syntax error if you eval the file on any other platform. Despite that you can't build it, you should be able to eval it.

Comment (malob) [reply pkgs/applications/misc/1password/default.nix]: @SuperSandro2000, good point, I think the changes I just pushed should address this.

**Comment by malob** [comment]
> I can't comment on the other comment, but could you add installing shell completions please? https://developer.1password.com/docs/cli/reference/commands/completion/

@maxeaubrey, done.

**Comment by malob** [comment]
Result of `nixpkgs-review pr 165321` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_1password</li>
  </ul>
</details>

**Comment by malob** [comment]
Result of `nixpkgs-review pr 165321` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_1password</li>
  </ul>
</details>

**Review by amaxine — APPROVED** [review]
Looks good, thank you for adding shell completions!

**Comment by malob** [comment]
Since this derivation overrides the `installPhase`, `postInstall` wasn't being run automatically. Just pushed a change that fixes that.

I now think this PR is ready to be merged.

**Comment by jdreaver** [comment]
I'm not opposed to merging, and I'm very excited for this, but this version of the CLI is not backwards-compatible with v1. How are breaking changes normally handled in `nixpkgs`? Should we create a new package like `_1password_2` to give folks the opportunity to migrate on their own time?

**Comment by malob** [comment]
@jdreaver, I had a look through the `nixpkgs` manual, and it doesn't seem to address this situation directly. The closest thing I found was that breaking changes shouldn't be backported to stable branches which makes sense.

Given my experience with Nix over the last few years my sense is that, users like myself that choose to use `{nixpkgs,nixos}-unstable` as their default channel/branch, get the benefit of having the latest versions of the packages, but that this is a double edged sword and sometimes there are costs, like new version of packages having big changes. Folks running on unstable branches can always overlay in the version of the package from the latest stable branch if they want to continue running the old version of a given package.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 165321` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>_1password</li>
  </ul>
</details>

File: pkgs/applications/misc/1password/default.nix [hunk 2]
```diff
@@ -32,25 +42,27 @@ stdenv.mkDerivation rec {
   '';
 
   installPhase = ''
-    install -D op $out/bin/op
+    install -D ${mainProgram} $out/bin/${mainProgram}
+    runHook postInstall
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/1password/default.nix:46]: Please also call preInstall

File: pkgs/applications/misc/1password/default.nix [hunk 3]
```diff
@@ -32,25 +42,27 @@ stdenv.mkDerivation rec {
...
   '';
 
-  dontStrip = stdenv.isDarwin;
+  postInstall = "installShellCompletion --cmd ${mainProgram}" + lib.concatMapStrings
+    (s: " --${s} <($out/bin/${mainProgram} completion ${s})") [ "bash" "fish" "zsh" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/1password/default.nix:50]: Please do it like this https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/cluster/argocd/default.nix#L50-L52

Comment (malob) [reply pkgs/applications/misc/1password/default.nix:50]: Looks like I hit the fancy limit :), will do next time I touch this package.

File: pkgs/applications/misc/1password/default.nix (hunk 2 above)
Comment (malob) [reply pkgs/applications/misc/1password/default.nix:46]: Will do next time I touch this package.

Comment (cole-h) [reply pkgs/applications/misc/1password/default.nix:46]: Done in https://github.com/NixOS/nixpkgs/pull/166339.

File: pkgs/applications/misc/1password/default.nix (hunk 3 above)
Comment (cole-h) [reply pkgs/applications/misc/1password/default.nix:50]: Also done in https://github.com/NixOS/nixpkgs/pull/166339.

**Pull request merged**